internal/sync2/cycle.go

```diff
@@ -0,0 +1,166 @@
+func (cycle *Cycle) SetInterval(interval time.Duration) {
+	cycle.interval = interval
+}
+
+func (cycle *Cycle) initialize() {
+	cycle.init.Do(func() {
+		cycle.stop = make(chan struct{})
+		cycle.control = make(chan interface{})
+	})
+}
+
+// Start runs the specified function with an errgroup
+func (cycle *Cycle) Start(ctx context.Context, group *errgroup.Group, fn func(ctx context.Context) error) {
+	group.Go(func() error {
+		return cycle.Run(ctx, fn)
+	})
+}
+
+// Run runs the specified function.
+func (cycle *Cycle) Run(ctx context.Context, fn func(ctx context.Context) error) error {
+	cycle.initialize()
+	defer close(cycle.stop)
+
+	currentInterval := cycle.interval
+	cycle.ticker = time.NewTicker(currentInterval)
+	if err := fn(ctx); err != nil {
+		return err
+	}
+	for {
+		select {
+
+		case message := <-cycle.control:
+			// handle control messages
+
+			switch message := message.(type) {
+			case cycleStop:
+				return nil
+
+			case cycleChangeInterval:
+				currentInterval = message.Interval
+				cycle.ticker.Stop()
+				cycle.ticker = time.NewTicker(currentInterval)
+
+			case cyclePause:
+				cycle.ticker.Stop()
+				// ensure we don't have ticks left
+				select {
+				case <-cycle.ticker.C:
+				default:
+				}
+
+			case cycleContinue:
+				cycle.ticker.Stop()
+				cycle.ticker = time.NewTicker(currentInterval)
+
+			case cycleTrigger:
+				// trigger the function
+				if err := fn(ctx); err != nil {
+					return err
+				}
+				if message.done != nil {
+					message.done <- struct{}{}
+				}
+			}
+
+		case <-ctx.Done():
+			// handle control messages
+			return ctx.Err()
+
+		case <-cycle.ticker.C:
+			// trigger the function
+			if err := fn(ctx); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// sendControl sends a control message
+func (cycle *Cycle) sendControl(message interface{}) {
+	cycle.initialize()
+	select {
+	case cycle.control <- message:
+	case <-cycle.stop:
+	}
+}
+
+// Stop stops the cycle permanently
+func (cycle *Cycle) Stop() {
+	cycle.sendControl(cycleStop{})
+}
+
+// ChangeInterval allows to change the ticker interval after it has started.
+func (cycle *Cycle) ChangeInterval(interval time.Duration) {
+	cycle.sendControl(cycleChangeInterval{interval})
+}
+
+// Pause pauses the cycle.
+func (cycle *Cycle) Pause() {
+	cycle.sendControl(cyclePause{})
+}
+
+// Restart restarts the ticker from 0.
+func (cycle *Cycle) Restart() {
+	cycle.sendControl(cycleContinue{})
+}
+
+// Trigger ensures that the loop is done at least once.
+// If it's currently running it waits for the previous to complete and then runs.
+func (cycle *Cycle) Trigger() {
+	cycle.sendControl(cycleTrigger{})
+}
+
+// TriggerWait ensures that the loop is done at least once and waits for completion.
+// If it's currently running it waits for the previous to complete and then runs.
+func (cycle *Cycle) TriggerWait() {
+	done := make(chan struct{})
+	cycle.sendControl(cycleTrigger{done})
+	select {
+	case <-done:
+	case <-cycle.stop:
+	}
```

Comment:
`close(done)`? I know normally it's best for writers to close channels, but it's hard in this case to know if the writer will ever even know about it, and once this function exits for either reason, we do know nothing else should want to write to it.

Reply:
added a close here